neurolib/models/multimodel/builder/adex.py

```diff
@@ -0,0 +1,944 @@
+        + cascade_table[y_idx + 1, x_idx] * (1 - dx_idx) * dy_idx
+        + cascade_table[y_idx + 1, x_idx + 1] * dx_idx * dy_idx
+    )
+
+
+class AdExMass(NeuralMass):
+    """
+    Adaptive exponential integrate-and-fire mean-field mass. Can be excitatory
+    or inhibitory, depending on the parameters.
+    """
+
+    name = "AdEx mean-field mass"
+    label = "AdExMass"
+    # define python callback function name for table lookup (linear-nonlinear
+    # approximation of Fokker-Planck equation)
+    python_callbacks = ["firing_rate_lookup", "voltage_lookup", "tau_lookup"]
+
+    num_noise_variables = 1
+
+    @staticmethod
+    def _rescale_strengths(params):
+        """
+        Rescale connection strengths for AdEx.
+        """
+        params = deepcopy(params)
+        assert isinstance(params, dict)
+        params["c_global"] = params["c_global"] * params["tau_syn_exc"] / params["J_exc_max"]
+
+        params["tau_m"] = params["C_m"] / params["g_L"]
+        return params
+
+    def __init__(self, params, lin_nonlin_cascade_filename=None, seed=None):
+        """
+        :param lin_nonlin_cascade_filename: filename for precomputed
+            linear-nonlinear cascade for AdEx, if None, will look for it in this
+            directory
+        :type lin_nonlin_cascade_filename: str|None
+        :param seed: seed for random number generator
+        :type seed: int|None
+        """
+        params = self._rescale_strengths(params)
+        super().__init__(params=params, seed=seed)
+        # use the same file as neurolib's native
+        lin_nonlin_cascade_filename = lin_nonlin_cascade_filename or os.path.join(
+            os.path.dirname(os.path.realpath(__file__)), "..", "..", "aln", "aln-precalc", DEFAULT_CASCADE_FILENAME,
+        )
+        self._load_lin_nonlin_cascade(lin_nonlin_cascade_filename)
+
+    def _load_lin_nonlin_cascade(self, filename):
+        """
+        Load precomputed linear-nonlinear cascade from h5 file.
+        """
+        # load linear-nonlinear cascade quantities from file
+        logging.info(f"Reading precomputed quantities from {filename}")
+        loaded_h5 = File(filename, "r")
+        self.mu_range = np.array(loaded_h5["mu_vals"])
+        self.d_mu = self.mu_range[1] - self.mu_range[0]
+        self.sigma_range = np.array(loaded_h5["sigma_vals"])
+        self.d_sigma = self.sigma_range[1] - self.sigma_range[0]
+        self.firing_rate_cascade = np.array(loaded_h5["r_ss"])
+        self.voltage_cascade = np.array(loaded_h5["V_mean_ss"])
+        self.tau_cascade = np.array(loaded_h5["tau_mu_exp"])
+        logging.info("Loading done, all quantities loaded.")
+        # close the file
+        loaded_h5.close()
+        self.lin_nonlin_fname = filename
+
+    def update_params(self, params_dict):
+        """
+        Update parameters as in the base class but also rescale.
+        """
+        # if we are changing C_m or g_L, update tau_m as well
```

Comment:
because of the rescaling it's messy to update parameters, but should be readable... practically all parameters are updated using `dict.update()` so here I'm just checking whether parameters from which the rescaling is done are also updated and if so, rescaling is done again